components/brave_wallet/browser/simulation_request_helper.cc

```diff
@@ -0,0 +1,195 @@
+/* Copyright (c) 2023 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include <utility>
+#include <vector>
+
+#include "base/base64.h"
+#include "base/strings/string_number_conversions.h"
+#include "base/values.h"
+#include "brave/components/brave_wallet/browser/json_rpc_requests_helper.h"
+#include "brave/components/brave_wallet/browser/rlp_encode.h"
+#include "brave/components/brave_wallet/browser/simulation_request_helper.h"
+#include "brave/components/brave_wallet/browser/solana_transaction.h"
+#include "brave/components/brave_wallet/common/brave_wallet_constants.h"
+#include "brave/components/brave_wallet/common/hex_utils.h"
+#include "brave/components/brave_wallet/common/solana_utils.h"
+
+namespace brave_wallet {
+
+base::Value::Dict GetMetadata(mojom::OriginInfoPtr origin_info) {
+  base::Value::Dict metadata_object;
+
+  if (origin_info && origin_info->origin_spec != "chrome://wallet" &&
+      origin_info->origin_spec != "brave://wallet") {
+    metadata_object.Set("origin", origin_info->origin_spec);
+  } else {
+    // TODO(onyb): We use "https://brave.com" as the default origin for now.
+    //  This is because Blowfish doesn't support "chrome://wallet" and
+    //  "brave://wallet" as origins yet. We'll update this once Blowfish
+    //  supports them.
+    metadata_object.Set("origin", "https://brave.com");
+  }
+
+  return metadata_object;
```

Comment:
Why this service needs transaction origin?
Could that lead to privacy leak when user makes a transaction with origin https://example.com and we send that origin to service?

Reply:
This is done to detect scam domains, and will be disclosed in our privacy policy. See https://github.com/brave/security/issues/1105.

Reply:
Technically we should not trust serialized origin coming from renderer. Only origins instances created in browser process could be used for such decision making. Not sure if this concern is applicable here btw

Reply:
Since we already _trust_ the origin as displayed in the panel during transaction confirmation, I don't see this any worse than the status quo. But tagging @kdenhartog to opine on this.

Reply:
I mean origin sent from browser process to renderer and back to browser is not OK. That should be a request from renderer to create a pending(or sort of) transaction object capturing current origin and using it for metadata. But this might be too complex to implement for not much benefit tbh.

Reply:
I'm not sure of the trade offs here yet. Let me do some research and get back to you on this.

cc @fmarier @diracdeltas who might be able to get an answer more quickly

Reply:
Ok, I looked into this further and agree with @supermassive that we should just get the Origin from the browser process rather than passing the origin from the browser process to the renderer and then back to the browser. The main reason being that this brings in inherently different assumptions depending on the platform we're on (e.g. Desktop/iOS/Android) which makes the code less cross platform compatible or allows for the potential of platform specific vulnerabilities.

Luckily it looks like when @supermassive added support for OriginInfo being passed to the Renderer here, it looks like he made it easy for other browser process code to easily get the origin.

So, I think we could directly get this information from the browser process by calling https://github.com/brave/brave-core/blob/e4d08f1c6c537a97e53aad4a978a549f1e1934bd/browser/brave_wallet/brave_wallet_provider_delegate_impl.cc#L94 and passing the Origin into https://github.com/brave/brave-core/blob/e4d08f1c6c537a97e53aad4a978a549f1e1934bd/components/brave_wallet/browser/brave_wallet_utils.cc#L1585 to keep this all within the browser process.

Reply:
@onyb is right, we need to use the origin info stored in tx when the tx was CREATED. @kdenhartog we can't just fetch origin based on currently active one would be when we need it because like @onyb mentioned "mismatch can easily happen".
It is the same concept that we don't decided `chain id` for tx based on currently active one when needed, we have to use the one when the tx was created, otherwise we will end up sending the tx onto the wrong network.

Reply:
Ahh great point @onyb I hadn't considered the impact of that case. Ok, in that case because we cannot limit the issue at the backend layer I guess we'll have to make sure that the origin isn't able to be tampered with at the UI layer for each platform. Let's go forward with this and I'll note that down to make sure I pay attention to that when reviewing the UI layers for this work.